libutils/stack.c

```diff
@@ -0,0 +1,212 @@
+    void **data;                      /**< Internal array of elements.     */
+    int size;                         /**< Amount of elements in stack.    */
+    int capacity;                     /**< Current memory allocated.       */
+};
+
+static void DestroyRange(Stack *stack, int start, int end);
+static void ExpandIfNecessary(Stack *stack);
+
+Stack *StackNew(size_t initial_capacity, void (ItemDestroy) (void *item))
+{
+    Stack *stack = xmalloc(sizeof(Stack));
+
+    if (initial_capacity <= 0)
+    {
+        initial_capacity = DEFAULT_CAPACITY;
+    }
+
+    pthread_mutex_t lock = PTHREAD_ERRORCHECK_MUTEX_INITIALIZER_NP;
+    stack->lock = lock;
+
+    stack->capacity = initial_capacity;
+    stack->size = STACK_EMPTY;
+    stack->data = xcalloc(sizeof(void *), initial_capacity);
+    stack->ItemDestroy = ItemDestroy;
+
+    return stack;
+}
+
+void StackDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        ThreadLock(&stack->lock);
+        DestroyRange(stack, STACK_EMPTY, stack->size);
+        ThreadUnlock(&stack->lock);
+    }
+
+    StackSoftDestroy(stack);
+}
+
+void StackSoftDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        pthread_mutex_destroy(&stack->lock);
+
+        free(stack->data);
+        free(stack);
+    }
+}
+
+void *StackPop(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+
+    int size = stack->size;
+    void *item = NULL;
+
+    if (size > STACK_EMPTY)
+    {
+        size--;
+        item = stack->data[size];
+
+        stack->data[size] = NULL;
+        stack->size = size;
+    }
+
+    ThreadUnlock(&stack->lock);
+
+    return item;
+}
+
+void StackPush(Stack *stack, void *item)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+    ExpandIfNecessary(stack);
+
+    stack->data[stack->size++] = item;
+
+    ThreadUnlock(&stack->lock);
+}
+
+int StackCount(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+    int count = stack->size;
+    ThreadUnlock(&stack->lock);
+
+    return count;
+}
+
+int StackCapacity(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+    int capacity = stack->capacity;
+    ThreadUnlock(&stack->lock);
+
+    return capacity;
+}
+
+bool StackIsEmpty(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+    bool empty = stack->size == STACK_EMPTY;
+    ThreadUnlock(&stack->lock);
+
+    return empty;
+}
+
+Stack *StackCopy(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(&stack->lock);
+
+    Stack *new_stack = StackNew(stack->capacity, stack->ItemDestroy);
+    for (int i = 0; i < stack->size; i++)
+    {
+        new_stack->data[i] = stack->data[i];
+    }
+
+    new_stack->size = stack->size;
+
+    ThreadUnlock(&stack->lock);
+
+    return new_stack;
+}
+
+static void DestroyRange(Stack *stack, int start, int end)
+{
+    assert(stack != NULL);
```

Comment:
Maybe add comments that these static functions assume that lock is already acquired.